uber/site_sections/preregistration.py

```diff
@@ -98,6 +98,48 @@ def index(self, message=''):
     def dealer_registration(self, message=''):
         return self.form(badge_type=c.PSEUDO_DEALER_BADGE, message=message)
 
+    def redeem(self, session, message='', **params):
```

Comment:
To build on what I said earlier and what @kitsuta has also said, I think there's a different approach we should take here.

The current workflow for the attendee is currently:
- Attendee fills out the reg form.
- Before they pay, they fill out the "redeem coupon" form.
- At the time they fill out that form, the validity is checked.
- Attendee then pays.

There are some problems with this approach.  A better approach would probably be:
- Attendee fills out the prereg form.  During this time, they can enter coupon codes into the "Comments" field.
- A discount is applied based on that coupon code.  However, the coupon is not actually redeemed until they try to pay.  It's validity is checked both at that time and when the payment goes through and the attendee is saved.

This implementation would be a little trickier, but I think it would be better in the long run.  With that being said, I'm happy to defer to Dom on this, who's working more closely with you on this, and who has spent more time reviewing this.